feat(forms): redo the signal forms metadata API

mmalerba opened the pull request:
This PR makes a number of changes to the metadata API to address design
flaws in the previous API. Some of the changes include:

- Replaces the previous `MetadataKey` and `AggregateMetadataKey` with a
  single unified `MetadataKey` that is used for all metadata.
- The new `MetadataKey` is only defined for fields that explicitly set
  it in their schema logic
- All metadata now has reducer / aggregate behavior
- The new `MetadataKey` has an option to create a managed key which
  wraps the result of its computed aggregate into some other structure
  such as a `Resource` or `linkedSignal`
- There are now two APIs to create metadata keys
  - `createMetadataKey` for pure computed metadata
  - `createManagedMetadataKey` for metadata that manages its computation
    internally

leonsenft commented on packages/forms/signals/src/api/logic.ts:
```diff
@@ -157,85 +157,33 @@ export function validateTree<TValue, TPathKind extends PathKind = PathKind.Root>
-): void {
+  key: TKey,
+  logic: NoInfer<LogicFn<TValue, MetadataSetterType<TKey>, TPathKind>>,
+): TKey {
```
Since `key` is no longer optional, should this still return the key? Previously it let us save the creation call:

```ts
const KEY = metadata(p, logic);
// vs
const KEY = createKey();
metadata(p, KEY, logic);
```

But now we have to explicitly create the key anyways:

```ts
const KEY = metadata(p, createMetadataKey(), logic);
// vs
const KEY = createMetadataKey();
metadata(p, KEY, logic);
```

leonsenft commented on packages/forms/signals/test/node/resource.spec.ts:
```diff
@@ -77,6 +80,11 @@ describe('resources', () => {
+    // TODO: what is this test trying to demonstrate? notably this resource does *not* contribute to the pending state.
+    // Access the validation state to trigger resource creation.
+    // (Metadata is created lazily when first accessed)
+    f().valid();
```
Does accessing `errors()` not pull on the validation and thus metadata?

It seems like a bug that we'd need to check for a validity status to trigger resource creation and thus produce the errors?

mmalerba replied:
Also eager/lazy aside, I think the reason I added that TODO is because I realized that the test is trying to demonstrate using a separately defined resource with `validate` _not_ `validateAsync`, and I noticed that it does not contribute to the `pending` state of the field.

I think this test might have been our attempt to show that if you already have some pre-existing external resource you can use it in your validation via `validate`, but I'm not sure its really a good solution if it doesn't report the pending state

leonsenft reviewed (changes requested):
nit: Maybe my mental model is off, but a "computed" metadata key seems a more fitting than a "wrapped" one. What about `computedMetadataKey()` and naming `wrap` -> `compute`?

mmalerba commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -6,189 +6,288 @@
  */
-export class AggregateMetadataKey<TAcc, TItem> {
-  private brand!: [TAcc, TItem];
+export class MetadataKey<TGet, TSet, TAcc> {
```
Done, maybe `TRead`, `TWrite` would be better names?

alxhub commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -143,6 +148,16 @@ export function andMetadataKey(): AggregateMetadataKey<boolean, boolean> {
 }
 
+export function resourceMetadataKey<TParams, TResult>(
+  factory: (params: Signal<TParams | undefined>) => ResourceRef<TResult>,
```
We probably want a plain `Resource` here:

```suggestion
  factory: (params: Signal<TParams | undefined>) => Resource<TResult>,
```

`ResourceRef` gives powers to e.g. destroy the resource, which we don't want to expose to consumers.

mmalerba replied:
I removed this because it wasn't actually used anywhere (remnant from previous iteration of the change). I could add it back if we want it, but right now its just a matter of call `createWrappedMetadataKey((...) => resource(...))`

mmalerba commented on packages/forms/signals/src/field/metadata.ts:
```diff
@@ -20,27 +20,27 @@ export class FieldMetadataState {
+        const logic = this.node.logicNode.logic.getMetadata(key);
+        let result = computed(() => logic.compute(this.node.context)) as T;
+        if (key.wrap) {
+          result = untracked(() =>
```
are we good with just saying that any wrapped key will be created on construction of the field, and allow plain computed keys to remain lazy? or add an explicit bit for this on the `MetadataKey`?

alxhub replied:
Yes, I think that's a reasonable assumption. Kind of like how I brought up in the naming discussion on Slack - one of the key features of wrapped metadata beyond just the final wrapping step is that it participates in the field lifecycle - it's attached to the field injector, is created with the field, and can be destroyed when the field is destroyed.

alxhub commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -6,189 +6,287 @@
+      if (item === undefined) {
+        return acc;
+      }
+      return Math.min(acc, item);
```
<img width="220" height="260" alt="image" src="https://github.com/user-attachments/assets/2c7a15a8-5bc7-4a31-bbea-87773df76594" />

mmalerba commented on packages/forms/signals/test/node/api/metadata.spec.ts:
```diff
@@ -0,0 +1,33 @@
+
+describe('metadata', () => {
+  it('should reduce values with andMetadataReducer', () => {
+    const KEY = createMetadataKey(andMetadataReducer());
```
yeah probably (though they're at least tested indirectly already through our built-in metadata keys)

leonsenft commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -0,0 +1,292 @@
+ * values to the same `MetadataKey` of the same field.
+ *
+ * @template TRead The type read from the `FieldState` for this key
+ * @template TWrite The written to this key using the `metadata()` rule
```
```suggestion
 * @template TWrite The type written to this key using the `metadata()` rule
```

leonsenft commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -0,0 +1,292 @@
+ * Creates a metadata key used to contain a computed value.
+ * The last value set on a given field tree node overrides any previously set values.
+ *
+ * @template TWrite The written to this key using the `metadata()` rule
```
```suggestion
 * @template TWrite The type written to this key using the `metadata()` rule
```

leonsenft commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -0,0 +1,292 @@
+ * Creates a metadata key used to contain a computed value.
+ *
+ * @param reducer The reducer used to combine individually set values into the final computed value.
+ * @template TWrite The written to this key using the `metadata()` rule
```
```suggestion
 * @template TWrite The type written to this key using the `metadata()` rule
```

leonsenft commented on packages/forms/signals/src/api/metadata.ts:
```diff
@@ -0,0 +1,292 @@
+ * @param wrap A function used to wrap the final computed value
+ *   (for example to create a resource, linkedSignal, etc out of the computed).
+ * @template TRead The type read from the `FieldState` for this key
+ * @template TWrite The written to this key using the `metadata()` rule
```
```suggestion
 * @template TWrite The type written to this key using the `metadata()` rule
```

leonsenft reviewed (approved):
approved-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

mmalerba commented on packages/forms/signals/src/api/rules/metadata.ts:
```diff
@@ -6,109 +6,98 @@
+  /** The reduce function. */
+  reduce: (acc: TAcc, item: TItem) => TAcc;
+  /** Gets the initial accumulated value. */
+  getInitial: NoInfer<() => TAcc>;
```
no difference that I can notice, removed. I think maybe in some previous iteration `getInitial` was optional and it mattered

mmalerba commented on packages/forms/signals/src/api/rules/metadata.ts:
```diff
@@ -6,109 +6,98 @@
-    (acc, item) => (item === undefined ? acc : [...acc, item]),
-    () => [],
-  );
+export function listMetadataReducer<TItem>(): MetadataReducer<TItem[], TItem | undefined> {
```
I do like the idea of the `Union` thing, though practically with `undefined` in the mix, the logic for a min/max is more complicated than just `Union(Math.min)`. I'll at least collapse the top-level symbols down under `MetadataReducer` though

mmalerba committed:
feat(forms): redo the signal forms metadata API

This PR makes a number of changes to the metadata API to address design
flaws in the previous API. Some of the changes include:

- Replaces the previous `MetadataKey` and `AggregateMetadataKey` with a
  single unified `MetadataKey` that is used for all metadata.
- The new `MetadataKey` is only defined for fields that explicitly set
  it in their schema logic
- All metadata now has reducer / aggregate behavior
- The new `MetadataKey` has an option to create a managed key which
  wraps the result of its computed aggregate into some other structure
  such as a `Resource` or `linkedSignal`
- There are now two APIs to create metadata keys
  - `createMetadataKey` for pure computed metadata
  - `createManagedMetadataKey` for metadata that manages its computation
    internally

mmalerba commented:
**caretaker:** adev test failure is unrelated
